### PR TITLE
CIN-02653: Text editor refinements

### PR DESCRIPTION
**Pull request by cysonya**
All changes pertain to the acceptance criteria in this ticket: https://cinchy.net/Apps/Integrated?appId=457&formId=905&rowId=2653

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -32,14 +32,18 @@
     "@fortawesome/free-solid-svg-icons": "^5.15.4",
     "@ng-bootstrap/ng-bootstrap": "^10.0.0",
     "@tiptap/core": "^2.0.0-beta.204",
+    "@tiptap/extension-code-block-lowlight": "^2.0.0-beta.209",
```
Comment (DAquilinaCinchy) [inline package.json:35]: This version should match the other TipTap extensions at 2.0.0-beta.204 if possible

File: src/app/dynamic-forms/enums/tiptap-mark-type.enum.ts [hunk 2]
```diff
@@ -92,7 +92,13 @@ export enum TiptapMarkType {
    */
   ListUnordered = "BULLET_LIST",
   /**
-   * Automatically inserts a <p> tag when doing a soft line breat
+   * Represents a <ul data-type="taskList">
```
Comment (DAquilinaCinchy) [inline src/app/dynamic-forms/enums/tiptap-mark-type.enum.ts:95]: I would rephrase this to `Represents a <ul> tag with the data-type attribute set to "task-list"` or something like that

File: src/app/dynamic-forms/fields/rich-text/rich-text.component.scss [hunk 3]
```diff
@@ -2,7 +2,9 @@
   padding: 0 !important;
 
   .editor-wrapper {
-    overflow: hidden;
+    max-height: 600px;
+    min-height: 100px;
+    overflow-y: auto;
```
Comment (DAquilinaCinchy) [inline src/app/dynamic-forms/fields/rich-text/rich-text.component.scss:7]: I think you also need to specify that overflow-x is hidden

File: src/app/dynamic-forms/fields/rich-text/rich-text.component.ts [hunk 4]
```diff
@@ -80,35 +86,50 @@ export class RichTextComponent implements OnDestroy, AfterViewInit {
   activeMarks = {
     bold: false,
     code: false,
-    heading1: false,
-    heading2: false,
-    heading3: false,
-    heading4: false,
-    heading5: false,
+    codeBlock: false,
+    Heading1: false,
```
Comment (DAquilinaCinchy) [inline src/app/dynamic-forms/fields/rich-text/rich-text.component.ts:90]: Why were these capitalized?

File: src/app/dynamic-forms/fields/rich-text/rich-text.component.ts [hunk 5]
```diff
@@ -80,35 +86,50 @@ export class RichTextComponent implements OnDestroy, AfterViewInit {
...
     faListUl: faListUl,
     faStrikethrough: faStrikethrough,
+    faTasks: faTasks,
     faUnderline: faUnderline
   };
 
   tiptapMarkType = TiptapMarkType;
 
 
+
```
Comment (DAquilinaCinchy) [inline src/app/dynamic-forms/fields/rich-text/rich-text.component.ts:132]: This extra line should be removed

File: src/app/dynamic-forms/fields/rich-text/rich-text.component.ts [hunk 6]
```diff
@@ -214,6 +253,13 @@ export class RichTextComponent implements OnDestroy, AfterViewInit {
       case TiptapMarkType.Code:
         this.editor?.commands.toggleCode();
 
+        break;
+      case TiptapMarkType.CodeBlock:
+        this.editor?.commands.toggleCodeBlock();
+
+        break;
+      case TiptapMarkType.Paragraph:
```
Comment (DAquilinaCinchy) [inline src/app/dynamic-forms/fields/rich-text/rich-text.component.ts:261]: Is there an explicit paragraph function on this? I thought paragraphs were added automatically to every line break.

File: src/styles.scss [hunk 7]
```diff
@@ -348,4 +348,43 @@ app-search-dropdown {
  .tooltip-inner{
   max-width: 250;
   text-align: left;
+ }
+
+ .tooltip-newline {
```
Comment (DAquilinaCinchy) [inline src/styles.scss:353]: There's an extra space at the beginning of this line

File: src/styles.scss [hunk 8]
```diff
@@ -348,4 +348,43 @@ app-search-dropdown {
...
+  text-align: center;
+  white-space: pre-line;
+}
+
+ .ProseMirror {
```
Comment (DAquilinaCinchy) [inline src/styles.scss:358]: There's an extra space at the beginning of this line

File: src/app/dynamic-forms/fields/rich-text/rich-text.component.ts [hunk 9]
```diff
@@ -80,35 +86,50 @@ export class RichTextComponent implements OnDestroy, AfterViewInit {
...
     italic: false,
     link: false,
     listOrdered: false,
     listUnordered: false,
+    listTask: false,
     strike: false,
     underline: false
   };
+  // Keyboard shortcut label for ctrl key
+  ctrlLabel = window.navigator.appVersion.indexOf("Mac") !== -1 ? "⌘" : "^";
```
Comment (DAquilinaCinchy) [inline src/app/dynamic-forms/fields/rich-text/rich-text.component.ts:104]: This is very clever, but will people inherently know that ^ refers to ctrl?

**Review by DAquilinaCinchy — CHANGES_REQUESTED** [review]

File: src/app/dynamic-forms/fields/rich-text/rich-text.component.ts (hunk 4 above)
Comment (cysonya) [reply src/app/dynamic-forms/fields/rich-text/rich-text.component.ts:90]: In the headings dropdown, I'm looping through `headings` where the array contains `{ value: "Heading1"},`.   Then in the html element I'm using `[selected]="activeMarks[heading.value]"` and `[value]="tiptapMarkType[heading.value]"`. Which uses the `value`, so I was thinking of avoiding to do `{ selected: 'heading1', value: 'Heading1'}` for this loop. Let me know if I should revert :P

File: src/app/dynamic-forms/fields/rich-text/rich-text.component.ts (hunk 6 above)
Comment (cysonya) [reply src/app/dynamic-forms/fields/rich-text/rich-text.component.ts:261]: In the dropdown a 'Paragraph' command was added, so I needed this to revert a heading back to a paragraph.

File: src/app/dynamic-forms/fields/rich-text/rich-text.component.ts (hunk 4 above)
Comment (DAquilinaCinchy) [reply src/app/dynamic-forms/fields/rich-text/rich-text.component.ts:90]: I think the better plan would be to keep this set consistent, and then in the headings array just set the value to "heading#" instead. Is there any reason why the value of the selected headings element would need to be capitalized?

File: src/app/dynamic-forms/fields/rich-text/rich-text.component.ts (hunk 6 above)
Comment (DAquilinaCinchy) [reply src/app/dynamic-forms/fields/rich-text/rich-text.component.ts:261]: Fair enough!

File: src/app/dynamic-forms/fields/rich-text/rich-text.component.ts (hunk 4 above)
Comment (cysonya) [reply src/app/dynamic-forms/fields/rich-text/rich-text.component.ts:110]: The reason the selected headings need to be capitalized is because on select change, it triggers `toggleMark` which requires `Heading#`.  Should I just add another key/value pair to keep this consistent then?

File: src/app/dynamic-forms/fields/rich-text/rich-text.component.ts (hunk 9 above)
Comment (DAquilinaCinchy) [reply src/app/dynamic-forms/fields/rich-text/rich-text.component.ts:126]: Works for me

File: src/app/dynamic-forms/fields/rich-text/rich-text.component.ts (hunk 4 above)
Comment (DAquilinaCinchy) [reply src/app/dynamic-forms/fields/rich-text/rich-text.component.ts:110]: Yeah, I don't like that. If we end up changing the structure of the enum then this will break and it will be really difficult to debug. I think in the headings array you should use something like `mark: TipTapMarkType.Heading#` and then use `[value]="heading.mark"` in the template. For the active flag on the list items, create a proxy function that takes in the mark type and returns the correct activeMark from a switch case. It's more verbose, for sure, but it's 100% clear and later will be much easier to unit test.

File: src/app/dynamic-forms/fields/rich-text/rich-text.component.ts [hunk 10]
```diff
@@ -156,22 +179,51 @@ export class RichTextComponent implements AfterViewInit, OnDestroy {
...
+            lowlight,
+          }),
           Link.extend({
+            addKeyboardShortcuts() {
+              return {
+                'Mod-k': () => {
+                  self.toggleLink()
+                  return true                
+                },
+                'Ctrl-k': () => {
```
Comment (DAquilinaCinchy) [inline src/app/dynamic-forms/fields/rich-text/rich-text.component.ts:214]: Using `Mod` is a shorthand for `Cmd` on Mac or `Ctrl` otherwise, so this second block is unnecessary.

File: src/app/dynamic-forms/fields/rich-text/rich-text.component.ts [hunk 11]
```diff
@@ -156,22 +179,51 @@ export class RichTextComponent implements AfterViewInit, OnDestroy {
...
       content = this.field.value;
     }
-
+    
     if (this.canEdit) {
+      // Disable spellcheck in code blocks
+      const CustomCodeBlockLowlight = CodeBlockLowlight.extend({
+        addAttributes() {
+          return {
+            spellcheck: { default: 'false' },
```
Comment (DAquilinaCinchy) [inline src/app/dynamic-forms/fields/rich-text/rich-text.component.ts:196]: Please use double quotes for strings or back-ticks (where appropriate) for strings

**Review by DAquilinaCinchy — CHANGES_REQUESTED** [review]

**Review by DAquilinaCinchy — APPROVED** [review]

**Review by cinchy-m — APPROVED** [review]
